### PR TITLE
feat: set up, read, and visualize job queue

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

Add support for reporting job queue and job status for all active jobs.
Check out where your job is in the queue and what status it has through
Determined WebUI, and CLI.

Part one of the queue management project.
Co-authored with @eecsliu

https://github.com/determined-ai/determined/pull/3231

- Backend & APIs:
  - define and persist jobs (landed)
  - refactor config for consistent resolution of priority, weight, and preemption status for jobs
  - K8: TODO
  - compute & advertise job queues different resource managers.
  - advertise job status for experiments
- WebUI
  - add a new job queue page to render job queues.
  - augment experiment state with job status in experiment details.
  - generic action dropdown (landed)
- CLI:
  - cli code gen based on openapi spec (separated)
  - query and display job queue in various formats

## Commentary (optional)

this splits the read portion from the [job-queue branch](https://github.com/determined-ai/determined/tree/job-queue)
code changes up to 6ae49418a80997360126f8c6d2cdabad1ba0c6e9 have been reviewed in prior separate PRs listed [here](https://github.com/determined-ai/determined/pulls?q=is%3Apr+author%3Ahamidzr+is%3Aclosed) and [here](https://github.com/determined-ai/determined/pulls?q=is%3Apr+is%3Aclosed+author%3Aeecsliu)

```
pick 0e937fcce feat: jq cli interface to read the queue (#3191)
pick 0b629cca4 feat: jq init read-only web ui for job queue (#3167)
pick 22d7ddd4c api: proto/ and subsequent generated code changes
pick b340e9705 master: WIP read changes
pick 327fa2d47 tmp: add to read
pick 6ae49418a read: remove sortTasksWithPosition usage
pick 9aeb98190 reuse job stats
d c8cd85c44 Revert "reuse job stats"
d 623bb1003 Revert "read: remove sortTasksWithPosition usage"
d 4d46474f8 Revert "tmp: add to read"
d 0c82d73c3 master: WIP write changes
d fbad23e15 Revert "Revert "reuse job stats""

# Rebase b41ae1dde..fbad23e15 onto b41ae1dde (12 commands)
#
```

## Checklist

- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [x] review new parts
  - [ ] new webui commits: @caleb
- [x] update api path?
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] mark or avoid advertising write operation APIs
- [x] docs?

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/themes.ts [hunk 1]
```diff
@@ -254,16 +254,20 @@ const stateColorMapping = {
   [SlotState.Free]: 'free',
   [SlotState.Pending]: 'pending',
   [SlotState.Running]: 'active',
+  [JobState.SCHEDULED]: 'active',
+  [JobState.SCHEDULEDBACKFILLED]: 'active',
+  [JobState.QUEUED]: 'suspended',
 };
 
-type States = RunState | CommandState | ResourceState | CheckpointState | SlotState
+export type StateOfUnion = RunState | CommandState | ResourceState | CheckpointState |
```
Comment (ioga) [inline webui/react/src/themes.ts:262]: variable naming 10/10

**Comment by hamidzr** [comment]
http://ec2-34-209-68-8.us-west-2.compute.amazonaws.com:8080

File: master/internal/command/api.go [hunk 2]
```diff
@@ -13,17 +14,22 @@ func RegisterAPIHandler(
 	system *actor.System,
 	echo *echo.Echo,
 	db *db.PgDB,
+	mConfig *config.Config,
 	middleware ...echo.MiddlewareFunc,
 ) {
-	system.ActorOf(actor.Addr("commands"), &commandManager{db: db})
+	system.ActorOf(actor.Addr("commands"),
+		&commandManager{db: db, mConfig: mConfig})
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/command/api.go:21]: yeah, that sounds like a good idea. let me take a look at that

[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/command/api.go:21]: a change at this scale seems good to me to do as part of this work. https://github.com/determined-ai/determined/compare/jq-read...hamidzr:config-singleton maybe there are many other references to m.config we could probably visit but that'd add to the surface area here

File: master/internal/api_job.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+			return false
+		}
+		if jobs[i].Summary.JobsAhead < jobs[j].Summary.JobsAhead {
+			return true
+		}
+		if jobs[i].Summary.JobsAhead > jobs[j].Summary.JobsAhead {
+			return false
+		}
+		return jobs[i].JobId < jobs[j].JobId
+	})
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api_job.go:44]: both works. IMO sorting fits nicely with pagination in this API layer unless we pass it on to jobs actor with pagination to potentially break out of some situations early

File: master/internal/command/api.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/command/api.go:21]: draft pr for this https://github.com/determined-ai/determined/pull/3277

File: prepqueue.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (hamidzr) [inline prepqueue.sh:1]: remove

File: master/internal/trial.go [hunk 5]
```diff
@@ -217,7 +218,7 @@ func (t *trial) maybeAllocateTask(ctx *actor.Context) error {
 			SingleAgent: false,
 		},
 
-		Preemptible:  true,
+		Preemptible:  true, // TODO should this read from config files?
```
Comment (hamidzr) [inline master/internal/trial.go:221]: remove

File: master/internal/resourcemanagers/resource_managers.go [hunk 6]
```diff
@@ -60,3 +61,23 @@ func GetResourceManagerType(rmConfig *ResourceManagerConfig) string {
...
+	}
+	sort.Ints(keys)
+
+	for _, k := range keys {
+		out = append(out, taskMap[k]...)
+	}
+	return out
+}
+
+func trueFilter(req *sproto.AllocateRequest) bool {
```
Comment (hamidzr) [inline master/internal/resourcemanagers/resource_managers.go:81]: use a lambda?

**Review by eecsliu — APPROVED** [review]

**Comment by hamidzr** [comment]
we got some new merge conflicts just now. rebasing

**Comment by hamidzr** [comment]
we need admin privileges to merge. @eecsliu could you squash merge, please?

File: master/internal/command/api.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/command/api.go]: I see this now, and I have to say I do not really like the singletons. See my comments on #3330 .

**Pull request merged**